File: src/lib/upload_video.ts
import fs from "fs/promises";
import OAuth from "oauth-1.0a";
import { createHmac } from "crypto";

const MEDIA_ENDPOINT_URL = "https://upload.twitter.com/1.1/media/upload.json";

const oauth = new OAuth({
  consumer: {
    key: process.env.TWITTER_CONSUMER_KEY!.trim(),
    secret: process.env.TWITTER_CONSUMER_SECRET!.trim(),
  },
  signature_method: "HMAC-SHA1",
  hash_function(base_string, key) {
    return createHmac("sha1", key).update(base_string).digest("base64");
  },
});

async function uploadInit(totalBytes: number): Promise<string> {
  console.log("UPLOAD INIT", totalBytes, totalBytes.toString());
  const request_data = {
    url: MEDIA_ENDPOINT_URL,
    method: "POST",
    data: {
      command: "INIT",
      total_bytes: totalBytes.toString(),
      media_type: "video/mp4",
      media_category: "tweet_video",
    },
  };

  const authorization = oauth.authorize(request_data, {
    key: process.env.TWITTER_ACCESS_TOKEN!.trim(),
    secret: process.env.TWITTER_ACCESS_TOKEN_SECRET!.trim(),
  });

  const response = await fetch(MEDIA_ENDPOINT_URL, {
    method: "POST",
    headers: {
      ...oauth.toHeader(authorization),
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams(request_data.data),
  });

  const data = await response.json();
  // console.log(data);
  return data.media_id;
}

async function uploadAppend(
  mediaId: string,
  chunk: Buffer,
  segmentIndex: number
): Promise<boolean> {
  console.log("UPLOAD APPEND", mediaId, chunk.length, segmentIndex);
  const formData = new FormData();
  formData.append("command", "APPEND");
  formData.append("media_id", mediaId);
  formData.append("segment_index", segmentIndex.toString());
  formData.append("media", new Blob([chunk]));

  const request_data = {
    url: MEDIA_ENDPOINT_URL,
    method: "POST",
    data: formData,
  };

  const authorization = oauth.authorize(request_data, {
    key: process.env.TWITTER_ACCESS_TOKEN!.trim().trim(),
    secret: process.env.TWITTER_ACCESS_TOKEN_SECRET!.trim().trim(),
  });

  const response = await fetch(MEDIA_ENDPOINT_URL, {
    method: "POST",
    headers: {
      ...oauth.toHeader(authorization),
    },
    body: formData,
  });
  console.log(await response.text());
  return response.ok;
}

async function uploadFinalize(mediaId: string) {
  console.log("UPLOAD FINALIZE", mediaId);

  const request_data = {
    url: MEDIA_ENDPOINT_URL,
    method: "POST",
    data: {
      command: "FINALIZE",
      media_id: mediaId,
    },
  };

  const authorization = oauth.authorize(request_data, {
    key: process.env.TWITTER_ACCESS_TOKEN!.trim(),
    secret: process.env.TWITTER_ACCESS_TOKEN_SECRET!.trim(),
  });

  const response = await fetch(MEDIA_ENDPOINT_URL, {
    method: "POST",
    headers: {
      ...oauth.toHeader(authorization),
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams(request_data.data),
  });

  const data = await response.json();
  return data;
}

export async function uploadVideo(videoPath: string): Promise<string> {
  const fileStats = await fs.stat(videoPath);
  const totalBytes = fileStats.size;
  let bytesCount = 0;
  const mediaId = await uploadInit(totalBytes);

  const chunkSize = 4 * 1024 * 1024; // 4MB chunks
  let segmentIndex = 0;
  const fileHandle = await fs.open(videoPath, "r");
  for (let i = 0; i < totalBytes; i += chunkSize) {
    const { buffer, bytesRead } = await fileHandle.read(
      Buffer.alloc(chunkSize),
      0,
      chunkSize,
      i
    );
    bytesCount += bytesRead;
    console.log("READ ", bytesRead, " bytes ", bytesCount, "/", totalBytes);
    const appendResult = await uploadAppend(mediaId, buffer, segmentIndex++);
    if (!appendResult) {
      //TODO if I eventually find the bug,then implement failed chunk re-uploads
      throw new Error(`APPEND FAILED FOR CHUNK ${segmentIndex}. exiting...`);
    }
  }
  await fileHandle.close();
  const finalizeResult = await uploadFinalize(mediaId);
  console.log(
    "MEDIA ID is",
    mediaId,
    " and finalize result returned ",
    finalizeResult
  );

  return mediaId;
}

File: src/lib/upload_video.py
import os
import sys
import time
import requests
import argparse
import json
from requests_oauthlib import OAuth1

MEDIA_ENDPOINT_URL = 'https://upload.twitter.com/1.1/media/upload.json'

def parse_args():
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(description='Upload video to Twitter.')
    parser.add_argument('--consumer_key', required=True, help='Twitter API Consumer Key')
    parser.add_argument('--consumer_secret', required=True, help='Twitter API Consumer Secret')
    parser.add_argument('--access_token', required=True, help='Twitter API Access Token')
    parser.add_argument('--access_token_secret', required=True, help='Twitter API Access Token Secret')
    parser.add_argument('--video_filename', required=True, help='Path to the video file to upload')
    return parser.parse_args()

class VideoTweet:
    def __init__(self, file_name, oauth):
        """Initialize video tweet properties."""
        self.video_filename = file_name
        self.total_bytes = os.path.getsize(self.video_filename)
        self.media_id = None
        self.processing_info = None
        self.oauth = oauth
        self.successful = False
        self.logs = []  # Collect logs for the result JSON

    def log(self, message):
        """Add a message to the logs."""
        # print(message)  # Optional: print to console for real-time feedback
        self.logs.append(message)

    def upload_init(self):
        """Initialize the upload."""
        self.log('INIT')
        request_data = {
            'command': 'INIT',
            'media_type': 'video/mp4',
            'total_bytes': self.total_bytes,
            'media_category': 'tweet_video'
        }
        req = requests.post(url=MEDIA_ENDPOINT_URL, data=request_data, auth=self.oauth)
        self.media_id = req.json()['media_id']
        self.log(f'Media ID: {self.media_id}')

    def upload_append(self):
        """Upload media in chunks."""
        segment_id = 0
        bytes_sent = 0
        with open(self.video_filename, 'rb') as file:
            while bytes_sent < self.total_bytes:
                chunk = file.read(4 * 1024 * 1024)
                self.log('APPEND')
                request_data = {
                    'command': 'APPEND',
                    'media_id': self.media_id,
                    'segment_index': segment_id
                }
                files = {'media': chunk}
                req = requests.post(url=MEDIA_ENDPOINT_URL, data=request_data, files=files, auth=self.oauth)

                if not (200 <= req.status_code < 300):
                    self.log(f"Error {req.status_code}: {req.text}")
                    sys.exit(0)

                segment_id += 1
                bytes_sent = file.tell()
                self.log(f'{bytes_sent} of {self.total_bytes} bytes uploaded')

        self.log('Upload chunks complete.')

    def upload_finalize(self):
        """Finalize the upload."""
        self.log('FINALIZE')
        request_data = {'command': 'FINALIZE', 'media_id': self.media_id}
        req = requests.post(url=MEDIA_ENDPOINT_URL, data=request_data, auth=self.oauth)
        self.log(json.dumps(req.json(), indent=2))
        self.processing_info = req.json().get('processing_info', None)
        self.check_status()

    def check_status(self):
        """Check the processing status."""
        if not self.processing_info:
            return

        state = self.processing_info['state']
        self.log(f'Media processing status is {state}')

        if state == 'succeeded':
            self.successful = True
            return

        if state == 'failed':
            self.log('Processing failed.')
            sys.exit(0)

        check_after_secs = self.processing_info['check_after_secs']
        self.log(f'Checking after {check_after_secs} seconds')
        time.sleep(check_after_secs)
        self.request_status()

    def request_status(self):
        """Request the current processing status."""
        self.log('STATUS')
        request_params = {'command': 'STATUS', 'media_id': self.media_id}
        req = requests.get(url=MEDIA_ENDPOINT_URL, params=request_params, auth=self.oauth)
        self.processing_info = req.json().get('processing_info', None)
        self.check_status()

    def generate_result(self):
        """Generate the final result JSON object."""
        final_state = f'SUCCESS {self.media_id}' if self.successful else 'ERROR'
        result = {
            'logs': self.logs,
            'final_state': final_state
        }
        print(json.dumps(result, indent=2))  # Output the result as JSON

if __name__ == '__main__':
    args = parse_args()
    oauth = OAuth1(
        args.consumer_key,
        client_secret=args.consumer_secret,
        resource_owner_key=args.access_token,
        resource_owner_secret=args.access_token_secret
    )
    video_tweet = VideoTweet(args.video_filename, oauth)
    video_tweet.upload_init()
    video_tweet.upload_append()
    video_tweet.upload_finalize()
    video_tweet.generate_result()
